sdk/program/src/stake_history.rs

```diff
@@ -84,10 +87,133 @@ impl Deref for StakeHistory {
+    }
+}
+
+// required for SysvarCache
+impl StakeHistoryGetEntry for Arc<StakeHistory> {
+    fn get_entry(&self, epoch: Epoch) -> Option<StakeHistoryEntry> {
+        self.deref().get_entry(epoch)
+    }
+}
+
+// precompute so we can statically allocate buffer
+const EPOCH_AND_ENTRY_SERIALIZED_SIZE: u64 = 32;
+
+impl StakeHistoryGetEntry for StakeHistorySyscall {
+    fn get_entry(&self, target_epoch: Epoch) -> Option<StakeHistoryEntry> {
+        let current_epoch = self.0;
+        let newest_historical_epoch = current_epoch.checked_sub(1).unwrap(); // HANA irrefutable, but do it nicer
+        let oldest_historical_epoch = newest_historical_epoch.saturating_sub(MAX_ENTRIES as u64);
+
+        // target epoch is before the beginning of time; this is a user error
+        // HANA panic/result?
+        if target_epoch == 0 {
+            return None;
+        }
+
+        // target epoch is old enough to have fallen off history; presume fully active/deactive
+        if target_epoch < oldest_historical_epoch {
+            return None;
+        }
+
+        // epoch delta is how many epoch-entries we offset in the stake history vector, which may be zero
+        // None means target epoch is current or in the future; this is a user error
+        // HANA panic/result?
+        let epoch_delta = match newest_historical_epoch.checked_sub(target_epoch) {
+            Some(d) => d,
+            None => return None,
+        };
+
+        // offset is the number of bytes to our desired entry, including eight for vector length
+        let offset =
+            epoch_delta * EPOCH_AND_ENTRY_SERIALIZED_SIZE + std::mem::size_of::<u64>() as u64;
+
+        let id_addr = StakeHistory::id().0.as_ptr();
+        let mut entry_buf = [0; EPOCH_AND_ENTRY_SERIALIZED_SIZE as usize];
+        let entry_buf_addr = &mut entry_buf as *mut _ as *mut u8;
+
+        #[cfg(target_os = "solana")]
+        let result = unsafe {
+            crate::syscalls::sol_get_sysvar(
+                id_addr,
+                entry_buf_addr,
+                offset,
+                EPOCH_AND_ENTRY_SERIALIZED_SIZE,
+            )
+        };
+
+        #[cfg(not(target_os = "solana"))]
+        let result = crate::program_stubs::sol_get_sysvar(
+            id_addr,
+            entry_buf_addr,
+            offset,
+            EPOCH_AND_ENTRY_SERIALIZED_SIZE,
+        );
+
+        match result {
+            crate::entrypoint::SUCCESS => {
+                if let Ok((entry_epoch, entry)) =
+                    bincode::deserialize::<(Epoch, StakeHistoryEntry)>(&entry_buf)
+                {
+                    // HANA panic/result? this can only happen if our math was wrong or the vector somehow became sparse
+                    if entry_epoch != target_epoch {
+                        return None;
+                    }
+
+                    Some(entry)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
```

Comment:
@joncinque curious for your thoughts on whether its semantically reasonable to return `None` for error conditions like this

the existing `StakeHistory` is just a vector, which `get()` searches for a given epoch, and if the epoch isnt there, it returns `None`. the important consideration is `None` is a _success_ condition: the `StakeHistoryEntry` was not found, so the `Delegation` function assumes that the stake is fullly active or inactive

the issue here is the `None` returns inside the `match` are _failure_ conditions. they should probably never be hit. the specific cases are things like "stake history became sparse so the offsets are wrong" (probably impossible), "stake history is shorter than it should be unexpectedly" (probably impossible), "stake history is not loaded in syscar cache" (almost certainly impossible), things like that

but i just wanted to get a second opinion, do you think its acceptable to return `None` for these? the alternative is returning `Result<Option<_>, _>` and modifying the `Delegation` functions such as `activating_and_deactivating()` to handle that... which im wary of touching that logic, and also its not clear to me what we would do, other than abort execution or just behave as if we got `None`

(i realized after pushing this that the bailing returns at the start of `get_entry()` are all fine, because it does the same thing as `get()`, so the stuff inside the `match` is all that im concerned about)

Reply:
This is a tricky one, since there's two things going on.

One the one hand, the code clearly checks that you're within the right bounds, so in normal circumstances, it would be better to `assert!` since otherwise there would be a programmer error.

On the other hand, things can change in the runtime, so the syscall could possibly fail in the future if there's ever a breaking change on the sysvar. This is unlikely, but not impossible. In those cases, it would be better to give the program a chance to notice the problem and recover rather than crash right away.

`Result<Option<...>>` seems a bit overkill, so all that to say, I agree with this approach.

Reply:
well the good news is the errors for the syscall are specified in extreme detail, because of the whole simd process. it ended up being a bit strange though because there are a couple cases where the sycall fails successfully (returns a non-`SUCCESS` error code), namely reading past the end of account data, or the sysvar not being available/supported (ie it was never populated in the sysvar cache, but in theory this would never happen)

in other words i dont think an update to the sysvar can get around needing to update the syscall and its interface. but ok if youre happy with the `None` returns im happy too